File: src/mlpy/stdlib/math.ml
// @description: Mathematical operations and constants with capability-based security
// @capability: read:math_constants
// @capability: execute:calculations
// @version: 1.0.0

/**
 * ML Math Standard Library
 * Provides mathematical operations with security validation
 */

capability MathOperations {
    allow read "math_constants";
    allow execute "calculations";
}

// Mathematical constants
pi = 3.141592653589793;
e = 2.718281828459045;
tau = 6.283185307179586;

// Basic mathematical functions
function abs(x) {
    if (x < 0) {
        return -x;
    } else {
        return x;
    }
}

function max(a, b) {
    if (a > b) {
        return a;
    } else {
        return b;
    }
}

function min(a, b) {
    if (a < b) {
        return a;
    } else {
        return b;
    }
}

function pow(base, exponent) {
    // Simple power implementation for positive integer exponents
    if (exponent == 0) {
        return 1;
    }

    result = 1;
    i = 0;
    while (i < exponent) {
        result = result * base;
        i = i + 1;
    }

    return result;
}

function sqrt(x) {
    // Newton's method for square root
    if (x < 0) {
        return -1; // Error case
    }

    if (x == 0) {
        return 0;
    }

    guess = x / 2;
    i = 0;

    while (i < 10) { // 10 iterations should be enough for reasonable precision
        better_guess = (guess + x / guess) / 2;
        if (abs(better_guess - guess) < 1e-10) {
            break;
        }
        guess = better_guess;
        i = i + 1;
    }

    return guess;
}

function factorial(n) {
    if (n <= 1) {
        return 1;
    }

    result = 1;
    i = 2;
    while (i <= n) {
        result = result * i;
        i = i + 1;
    }

    return result;
}

function gcd(a, b) {
    // Greatest common divisor using Euclidean algorithm
    while (b != 0) {
        temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

// Trigonometric functions (simplified approximations)
function sin_approx(x) {
    // Taylor series approximation: sin(x) ≈ x - x³/6 + x⁵/120
    x2 = x * x;
    x3 = x * x2;
    x5 = x3 * x2;

    return x - x3/6 + x5/120;
}

function cos_approx(x) {
    // Taylor series approximation: cos(x) ≈ 1 - x²/2 + x⁴/24
    x2 = x * x;
    x4 = x2 * x2;

    return 1 - x2/2 + x4/24;
}

File: src/mlpy/stdlib/math.py
"""ML Math Standard Library - Python Implementation."""

import math as py_math
from typing import Any


class Math:
    """ML math operations implemented in Python."""

    # Mathematical constants
    pi = py_math.pi
    e = py_math.e

    @staticmethod
    def sqrt(x: float) -> float:
        """Square root function."""
        if x < 0:
            return 0  # Error case for ML compatibility
        return py_math.sqrt(x)

    @staticmethod
    def abs(x: float) -> float:
        """Absolute value function."""
        return abs(x)

    @staticmethod
    def sin(x: float) -> float:
        """Sine function."""
        return py_math.sin(x)

    @staticmethod
    def cos(x: float) -> float:
        """Cosine function."""
        return py_math.cos(x)

    @staticmethod
    def tan(x: float) -> float:
        """Tangent function."""
        return py_math.tan(x)

    @staticmethod
    def ln(x: float) -> float:
        """Natural logarithm."""
        if x <= 0:
            return -999  # Error case for ML compatibility
        return py_math.log(x)

    @staticmethod
    def log(x: float, base: float = 10) -> float:
        """Logarithm with specified base."""
        if x <= 0 or base <= 0 or base == 1:
            return -999  # Error case
        return py_math.log(x, base)

    @staticmethod
    def exp(x: float) -> float:
        """Exponential function (e^x)."""
        try:
            return py_math.exp(x)
        except OverflowError:
            return float('inf')

    @staticmethod
    def pow(x: float, y: float) -> float:
        """Power function (x^y)."""
        try:
            return pow(x, y)
        except (OverflowError, ZeroDivisionError):
            return float('inf') if x > 0 else 0

    @staticmethod
    def floor(x: float) -> int:
        """Floor function."""
        return py_math.floor(x)

    @staticmethod
    def ceil(x: float) -> int:
        """Ceiling function."""
        return py_math.ceil(x)

    @staticmethod
    def round(x: float) -> int:
        """Round to nearest integer."""
        return round(x)

    @staticmethod
    def min(a: float, b: float) -> float:
        """Minimum of two values."""
        return min(a, b)

    @staticmethod
    def max(a: float, b: float) -> float:
        """Maximum of two values."""
        return max(a, b)

    @staticmethod
    def random() -> float:
        """Random number between 0 and 1."""
        import random
        return random.random()

    @staticmethod
    def degToRad(degrees: float) -> float:
        """Convert degrees to radians."""
        return py_math.radians(degrees)

    @staticmethod
    def radToDeg(radians: float) -> float:
        """Convert radians to degrees."""
        return py_math.degrees(radians)


# Global math instance for ML programs
math = Math()

# Constants
pi = py_math.pi
e = py_math.e

# Helper functions for ML bridge
def sqrt_helper(x: float) -> float:
    """Helper function for square root."""
    return math.sqrt(x)

def abs_helper(x: float) -> float:
    """Helper function for absolute value."""
    return math.abs(x)

def min_helper(a: float, b: float) -> float:
    """Helper function for minimum."""
    return math.min(a, b)

def max_helper(a: float, b: float) -> float:
    """Helper function for maximum."""
    return math.max(a, b)